File: leetcode/p00200/p00200.go
package p00200

// Time Complexity: O(mn)
// Space Complexity: O(n)
func numIslands(grid [][]byte) int {
	// SC: O(n)
	q := CreateQueue(len(grid) * len(grid[0]))

	// traverse the whole grid -> TC: O(n)
	count := 0
	for row, columns := range grid {
		for col, val := range columns {
			if val == '1' {
				// traveres whole island
				bfs(&q, grid, row, col)
				count++
			}
		}
	}

	return count
}

func bfs(q *Queue, grid [][]byte, row, col int) {
	// represent the left, top, right, down direction of the cursor
	checkList := [][]int{{0, -1}, {-1, 0}, {0, 1}, {1, 0}}

	maxRow := len(grid) - 1
	maxCol := len(grid[0]) - 1

	q.Enqueue([]int{row, col})
	for !q.IsEmpty() {
		pos, _ := q.Dequeue()
		for _, offset := range checkList {
			y := pos[0] + offset[0]
			x := pos[1] + offset[1]
			if x >= 0 && x <= maxCol && y >= 0 && y <= maxRow && grid[y][x] == '1' {
				q.Enqueue([]int{y, x})
				grid[y][x] = '2'
			}
		}
	}
}

type Queue struct {
	queue    [][]int
	head     int
	tail     int
	size     int
	capacity int
}

func CreateQueue(capacity int) Queue {
	return Queue{
		queue:    make([][]int, capacity),
		head:     0,
		tail:     capacity - 1,
		size:     0,
		capacity: capacity,
	}
}

func (this *Queue) IsEmpty() bool {
	return this.size == 0
}

func (this *Queue) IsFull() bool {
	return this.size == this.capacity
}

func (this *Queue) Enqueue(position []int) bool {
	if this.IsFull() {
		return false
	}

	this.tail++
	if this.tail == this.capacity {
		this.tail = 0
	}

	this.queue[this.tail] = position
	this.size++

	return true
}

func (this *Queue) Dequeue() ([]int, bool) {
	if this.IsEmpty() {
		return nil, false
	}

	pos := this.queue[this.head]

	this.head++
	if this.head == this.capacity {
		this.head = 0
	}

	this.size--
	return pos, true
}

File: leetcode/p00200/p00200.py
from collections import deque
from typing import List

class Solution:
    # Time Complexity: O(m+n)
    # Space Complexity: O(m)
    def numIslands_bfs(self, grid: List[List[str]]) -> int:
        # traverse whole grid
        count = 0
        for row in range(len(grid)):
            for col in range(len(grid[0])):
                if grid[row][col] == "1":
                    # traveres whole island
                    self.bfs(grid, row, col)
                    count += 1

        return count

    # Time Complexity: O(m)
    # Space Complexity: O(m)
    def bfs(self, grid: List[List[str]], row: int, col: int):
        gridHeight = len(grid)-1
        gridWidth = len(grid[0])-1

        q = deque([(row, col)])
        while q:
            row, col = q.popleft()
            if 0 <= row <= gridHeight and 0 <= col <= gridWidth and grid[row][col] == "1":
                grid[row][col] = "2"
                q.extend([
                    (row-1, col),
                    (row+1, col),
                    (row, col-1),
                    (row, col+1)])
    
    # Time Complexity: O(m+n)
    # Space Complexity: O(m)
    def numIslands_dfs(self, grid: List[List[str]]) -> int:
        count = 0

        for row in range(len(grid)):
            for col in range(len(grid[0])):
                if grid[row][col] == "1":
                    self.dfs(grid, row, col)
                    count += 1
        
        return count

    def dfs(self, grid: List[List[str]], row: int, col: int):
        grid[row][col] = "2"

        if row-1 >= 0 and grid[row-1][col] == "1":
            self.dfs(grid, row-1, col)
        
        if col+1 <= len(grid[0])-1 and grid[row][col+1] == "1":
            self.dfs(grid, row, col+1)
        
        if row+1 <= len(grid)-1 and grid[row+1][col] == "1":
            self.dfs(grid, row+1, col)

        if col-1 >= 0 and grid[row][col-1] == "1":
            self.dfs(grid, row, col-1)

if __name__ == '__main__':
    s = Solution()

    assert s.numIslands_bfs([
        ["1","1","1","1","0"],
        ["1","1","0","1","0"],
        ["1","1","0","0","0"],
        ["0","0","0","0","0"]]) == 1

    assert s.numIslands_bfs([
        ["1","1","0","0","0"],
        ["1","1","0","0","0"],
        ["0","0","1","0","0"],
        ["0","0","0","1","1"]]) == 3

    assert s.numIslands_dfs([
        ["1","1","1","1","0"],
        ["1","1","0","1","0"],
        ["1","1","0","0","0"],
        ["0","0","0","0","0"]]) == 1

    assert s.numIslands_dfs([
        ["1","1","0","0","0"],
        ["1","1","0","0","0"],
        ["0","0","1","0","0"],
        ["0","0","0","1","1"]]) == 3
